client/src/loaders/PromiseLoader.js

```diff
@@ -24,18 +24,21 @@ class PromiseLoader extends React.Component {
   componentDidMount() {
     const {promiseFn} = this.props;
     promiseFn()
-      .then(this.onResolved)
-      .then(this.onRejected);
+      .then(result => {
+        if (result.length > 0) {
+          this.onResolved(result);
+        } else{
+          this.onRejected(result);
+        }
+      });
   }
 
   onResolved(resolve) {
-    const {dataP} = this.state;
-    this.setState({ dataP: {...dataP, resolve} });
+    this.setState({ dataP: promiseState(false, resolve, undefined) });
   }
 
   onRejected(reject) {
-    const {dataP} = this.state;
-    this.setState({ dataP: {...dataP, reject} });
+    this.setState({ dataP: promiseState(false, false, true) });
```

Comment:
I can't follow why this change (and the related change to `promiseState` and `onResolved` are needed, can you explain a bit more?

Reply:
Hmmm. so I actually did not understand the syntax
`this.setState({ dataP: {...dataP, reject} });`
but from what I understood, it went and updated one of the three states (isPending, resolve, reject) in `dataP`. The problem I saw was that, in the case of `onRejected(reject)`, when it went to update `dataP` it would not clear the `isPending:true` that was originally in `dataP`. We would end up having dataP look something like
`{
isPending: true,
resolve: undefined,
reject: (whatever came back from the promise)
}`

Reply:
Got it, this is a great explanation, thanks! 👍 

To clarify,
```
this.setState({ dataP: {
  ...dataP,
  reject
} })
```
 works just like you described - the `...` "destructures" each property in the dataP object as if you had written them each out manually.  So it's the same as:

```
this.setState({ dataP: {
  isPending: dataP.isPending,
  resolve: dataP.resolve,
  reject: dataP.reject,
  reject
} })
```
and then the `reject` variable that comes after essentially overrides the `dataP.reject` line.

And yes, this should do something like: `this.setState({ dataP: {...dataP, resject, isPending: false})` to also set the `isPending` value.  This was my bug with `PromiseLoader` so thanks for figuring this out and patching it up!